src/app/forms/components/defect/defect.component.ts

```diff
@@ -1,85 +1,216 @@
+  }
+
+  ngOnInit(): void {
+    const defectIndex = this.store.pipe(select(selectRouteParam('defectIndex')));
+    const defectRef = this.store.pipe(select(selectRouteParam('ref')));
+
+    this.store
+      .select(toEditOrNotToEdit)
+      .pipe(
+        withLatestFrom(defectIndex, defectRef),
+        takeUntil(this.onDestroy$),
+        filter(([testResult]) => !!testResult)
+      )
+      .subscribe(([testResult, defectIndex, defectRef]) => {
+        this.defects = testResult!.testTypes[0].defects;
+        this.vehicleType = testResult!.vehicleType;
+        this._defectsForm = (this.dfs.createForm(DefectsTpl, testResult) as CustomFormGroup).get(['testTypes', '0', 'defects']) as CustomFormArray;
+        if (defectIndex) {
+          this.index = Number(defectIndex!);
+          this.form = this._defectsForm.controls[this.index] as CustomFormGroup;
+          this.defect = this.defects![this.index];
+        } else if (defectRef) {
+          this.store
+            .select(selectByDeficiencyRef(defectRef, this.vehicleType))
+            .pipe(take(1))
+            .subscribe(([defect, item, deficiency]) => {
+              this.initializeDefect(defect as Defect, item as Item, deficiency as Deficiency);
+            });
         }
       });
+
+    this.store
+      .select(selectByImNumber(this.defect?.imNumber || NaN, this.vehicleType))
+      .pipe(
+        takeUntil(this.onDestroy$),
+        filter(d => !!d)
+      )
+      .subscribe(defectsTaxonomy => {
+        this.initializeInfoDictionary(defectsTaxonomy);
+      });
+  }
+
+  ngOnDestroy(): void {
+    this.onDestroy$.next(true);
+    this.onDestroy$.complete();
+  }
+
+  get isDangerous(): boolean {
+    return this.defect!.deficiencyCategory === 'dangerous';
+  }
+
+  get isAdvisory(): Boolean {
+    return this.defect!.deficiencyCategory === 'advisory';
+  }
+
+  handleSubmit() {
+    const errors: GlobalError[] = [];
+    DynamicFormService.updateValidity(this.form, errors);
+
+    if (errors.length > 0) {
+      this.errorService.setErrors(errors);
     }
+
+    if (this.form.invalid) {
+      return;
+    }
+
+    if (this.index || this.index === 0) {
+      this.store.dispatch(updateDefect({ defect: this.form.getCleanValue(this.form) as TestResultDefect, index: this.index }));
+    } else {
+      this.store.dispatch(createDefect({ defect: this.form.getCleanValue(this.form) as TestResultDefect }));
+    }
+
+    this.navigateBack();
   }
 
-  @Output() removeDefect = new EventEmitter<number>();
+  handleRemove() {
+    this.store.dispatch(removeDefect({ index: this.index }));
+    this.navigateBack();
+  }
 
-  info?: AdditionalInfoSection;
+  navigateBack() {
+    this.resultService.updateResultOfTest();
+    this.router.navigate(['../..'], { relativeTo: this.activatedRoute, queryParamsHandling: 'preserve' });
+  }
 
-  infoDictionary: Record<string, Array<FormNodeOption<any>>> = {};
+  togglePRS() {
```

Comment:
Seems the `togglePRS()` and `toggleProhibition()` methods share very similar code and could be changed to a `toggleDefectField` method to reduce code repetition

Reply:
Also there is no need to patch the value of the whole `defect` object into the form as it is known which field is being updated. Something like this would work to solve these issues:
```js
  toggleDefectField(field: keyof TestResultDefect) {
    if (!this.defect) {
      return;
    }
    this.defect = { ...this.defect, [field]: !this.defect[field] } as TestResultDefect;
    this._defectsForm?.controls[this.index ?? this._defectsForm.length - 1].get(field)?.patchValue(this.defect[field]);
  }
```

Obviously changes would have to be done in the html to pass the correct name of the field to the method